File: Sample.BlazorWasm.TextEditor/wwwroot/src/Font.ts
namespace Sample.BlazorWasm.TextEditor {
    /**
     * 
     */
    export class Font {
        private readonly _measureCharOffscreen = new Offscreen({ w: 1, h: 1 });

        private readonly _family: string;
        private readonly _height: number;

        public get family() { return this._family; }
        public get height() { return this._height; }

        /**
         * [DotNetInvokable]
         */
        public static create(family: string, height: number): JSObjectId {
            const font = new Font(family, height);
            return JSObjectReference.create(font).id;
        }

        /**
         * 
         */
        public constructor(family: string, height: number) {
            this._family = family;
            this._height = height;

            this._measureCharOffscreen.ctx2d.font = `normal ${this._height}pt '${this._family}'`;
            this._measureCharOffscreen.ctx2d.textBaseline = "top";
        }

        /**
         *
         */
        public measureChar(ch: string): CharSize {
            const width = this._measureCharOffscreen.ctx2d.measureText(ch).width!;
            const heightInPixels = this._height / 72.0 * 96.0;
            return {
                width: width,
                height: heightInPixels,
            };
        }

        /**
         *
         */
        public measureTextWidth(text: string): number {
            return this._measureCharOffscreen.ctx2d.measureText(text).width!;
        }
    }
}

File: Sample.BlazorWasm.TextEditor/UIContextImpl/Font.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.Json;
using Crash.Core;
using Crash.Core.Collections;
using Crash.Core.Text;
using Crash.Core.UI.Common;
using Crash.Core.UI.UIContext;
using Microsoft.JSInterop;
using Sample.BlazorWasm.TextEditor.Common.Interop;

namespace Sample.BlazorWasm.TextEditor.UIContextImpl
{
    /// <summary>
    /// 
    /// </summary>
    [Immutable]
    public sealed class Font : IFont, IAutoFieldDisposable
    {
        private readonly Dictionary<char, CharSize> _measureCache = new Dictionary<char, CharSize>();

        [DisposableField]
        private readonly JSObjectReference _jsFont;

        public JSObjectReference JsObj => _jsFont;

        /// <summary>
        /// コンストラクタ。
        /// </summary>
        public Font(string family, double height, IJSInProcessRuntime jsRuntime)
        {
            _jsFont = JSObjectReference.From(jsRuntime,
                "Sample.BlazorWasm.TextEditor.Font.create", family, height);
        }

        /// <summary>
        /// 
        /// </summary>
        public CharSize MeasureChar(char ch)
        {
            return _measureCache.GetOrAdd(ch, (key, self) => self.InternalMeasureChar(key), this);
        }

        /// <summary>
        /// 
        /// </summary>
        private CharSize InternalMeasureChar(char ch)
        {
            Debug.Assert(CharUtil.CanDraw(ch));

            var size = _jsFont.Call<JsonElement>("measureChar", ch);
            return new CharSize(
                size.GetProperty("width").GetDouble(),
                size.GetProperty("height").GetDouble());
        }
    }
}
